src/ejabberd_c2s.erl

```diff
@@ -412,6 +421,13 @@ mechanism(S) ->
     #xmlel{name = <<"mechanism">>,
            children = [#xmlcdata{content = S}]}.
 
+filter_mechanism(<<"EXTERNAL">>, S) ->
+    case get_peer_cert(S) of
```

Comment:
Do we need to call `get_peer_cert` more than once?

Reply:
it allows to keep code cleaner and avoid extra parameter in ejabberd_c2s state, is it a problem?
if certificate is missing (empty cert) or invalid it should be proposed to the user to other SASL mechanisms.

Reply:
I mean we call it at least twice to auth a user.
Can we measure, that calling this function is cheap?
Otherwise we can keep cert in state, first time we call the function, until we actually don't need it (premature optimization here, but it can be in a code comment at least).

Reply:
:+1: for adding as a TODO in a comment

Reply:
1) premature optimization is evil
2) it's not possible to implement simple caching of certificate inside the tls socket record because of these reasons:
    * the only function that returns tls socket is **ejabbberd_tls:tls_to_tcp/2**, so we can cache the certificate only inside this function
    * for **just_tls** we can simply grab peer certificate right after **just_tls:tls_to_tcp/2**
    * but **fast_tls** doesn't perform TLS handshake on **fast_tls:tls_to_tcp/2** (handshake is accomplished after the first **fast_tls:recv_data/2** call) so obviously we cannot get any peer certificate after that call
3) to eliminate any potential drawbacks for non SASL External scenarios we can ensure that certificate is requested from tls backend (**just_tls**/**fast_tls**) only when it has sense:
    * for TLS client connection (server's certificate is mandatory for TLS handshake)
    * for TLS server  connection, when client's certificate is explicitly requested during the TLS handshake